src/engine/sys/sdl_glimp.cpp

```diff
@@ -558,196 +774,542 @@ static rserr_t GLimp_SetMode( int mode, bool fullscreen, bool noborder )
+			{
+				logger.Warn( "OpenGL 2.0 is not supported, trying 2.1 instead" );
 
-		if ( alphaBits < 0 )
-		{
-			alphaBits = 0;
+				customConfiguration.minor = 1;
+			}
 		}
 
-		depthBits = r_depthbits->integer;
+		logger.Debug( "GL version %d.%d is forced by r_MajorVersion and r_MinorVersion",
+			customConfiguration.major,
+			customConfiguration.minor );
+
+		customOptions = true;
+	}
+
+	if ( customConfiguration.profile == undefinedProfile )
+	{
+		customConfiguration.profile = bestConfiguration.profile;
+	}
+
+	customConfiguration.colorBits = std::max( 0, r_colorbits->integer );
 
-		if ( !depthBits )
+	if ( customConfiguration.colorBits == 0 )
+	{
+		customConfiguration.colorBits = bestConfiguration.colorBits;
+	}
+	else
+	{
+		if ( customConfiguration.colorBits != bestConfiguration.colorBits )
 		{
-			depthBits = 24;
+			logger.Debug( "Color framebuffer bitness %d is forced by r_colorbits",
+				customConfiguration.colorBits );
+
+			customOptions = true;
 		}
+	}
 
-		stencilBits = r_stencilbits->integer;
-		samples = r_ext_multisample->integer;
+	if ( customOptions )
+	{
+		if ( !GLimp_RecreateWindowWhenChange( fullscreen, bordered, customConfiguration ) )
+		{
+			return rserr_t::RSERR_INVALID_MODE;
+		}
+
+		const char* customProfileName = GLimp_getProfileName( customConfiguration.profile );
+
+		/* When calling GLimp_CreateContext() some drivers may
+		provide a valid context that is unusable while GL_CheckErrors()
+		catches nothing.
 
-		for ( i = 0; i < 4; i++ )
+		We can catch some errors when calling GLimp_DetectAvailableModes(),
+		see below. */
+
+		if ( GLimp_CreateContext( customConfiguration ) )
 		{
-			int testColorBits, testCore;
-			int major = r_glMajorVersion->integer;
-			int minor = r_glMinorVersion->integer;
+			logger.Notice( "Using custom context: %d-bit OpenGL %d.%d %s",
+				customConfiguration.colorBits,
+				customConfiguration.major,
+				customConfiguration.minor,
+				customProfileName );
 
-			// 0 - 24 bit color, core
-			// 1 - 24 bit color, compat
-			// 2 - 16 bit color, core
-			// 3 - 16 bit color, compat
-			testColorBits = (i >= 2) ? 16 : 24;
-			testCore = ((i & 1) == 0);
+			requestedConfiguration = customConfiguration;
+		}
+		else
+		{
+			logger.Warn( "Failed custom context: %d-bit OpenGL %d.%d %s",
+				customConfiguration.colorBits,
+				customConfiguration.major,
+				customConfiguration.minor,
+				customProfileName );
 
-			if( testCore && !Q_stricmp(r_glProfile->string, "compat") )
-				continue;
+			logger.Warn( "SDL_GL_CreateContext failed: %s", SDL_GetError() );
 
-			if( testCore && GLEWmajor < 2 )
-				continue;
+			customOptions = false;
+		}
+	}
 
-			if( !testCore && !Q_stricmp(r_glProfile->string, "core") )
-				continue;
+	return rserr_t::RSERR_OK;
+}
 
-			if( testColorBits > colorBits )
-				continue;
+static rserr_t GLimp_ApplyPreferredOptions( const bool fullscreen, const bool bordered, const glConfiguration &bestConfiguration, glConfiguration &requestedConfiguration )
```

Comment:
What is the point of this function?

Reply:
It is to select OpenGL 3.2 if a greater OpenGL version is validated when “extended validation” is enabled. When “extended validation” is disabled, if 3.2 is validated this would just do nothing and keep 3.2.

I remember some talk we had when you said there using greater version wasn't doing anything useful, and you probably also said it was better to only use what we need, which I agree with.

Hence I wrote this function to force using a 3.2 context if any greater core version is validated.

When extended validation is enabled, that would add at least one context recreation to downgrade from higher validated version (example, 4.6) to 3.2. With default behaviour, the engine would test 3.2 first, validate 3.2, but then when this function is called nothing will be done since context would already be 3.2 and then context will be kept as is.